include/libp2p/connection/stream_and_protocol.hpp

```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef LIBP2P_CONNECTION_STREAM_AND_PROTOCOL_HPP
+#define LIBP2P_CONNECTION_STREAM_AND_PROTOCOL_HPP
+
+#include <functional>
+#include <memory>
+
+#include <libp2p/outcome/outcome.hpp>
+#include <libp2p/peer/protocol.hpp>
+
+namespace libp2p::connection {
+  class Stream;
+}  // namespace libp2p::connection
+
+namespace libp2p {
+  /**
+   * Extends stream pointer with protocol.
+   * Used by `Host::newStream`, `Dialer::newStream`, `Host::setProtocolHandler`,
+   * `Router::setProtocolHandler`. Keeps `newStream` and `setProtocolHandler`
+   * callbacks simple and compatible.
+   */
+  struct StreamAndProtocol : std::shared_ptr<connection::Stream> {
```

Comment:
I kinda dislike that it's tied to a shared_ptr this way, I think it's done for compatibility but it appears to me that libp2p code base is not that big so that it can become a major problem.